File: Java/283-MoveZeroes.java
public class Solution {
    public void moveZeroes(int[] nums) {
        if (nums == null || nums.length == 0){
            return;
        }
        int len = nums.length, insertPoint = 0;
        for (int num : nums){
            if (num != 0){
                nums[insertPoint] = num;
                insertPoint++;
            }
        }
        for (int i = insertPoint; i < len; i++){
            nums[i] = 0;
        }
    }
}

//sol with two pointers
public class Solution {
    public void moveZeroes(int[] nums) {
        if (nums == null || nums.length == 0 || nums.length == 1){
            return;
        }
        int slow = 0, fast = 0;
        while (fast < nums.length){
            if (nums[fast] != 0){
                nums[slow] = nums[fast];
                slow++;
                fast++;
            }else{
                fast++;
            }
        }
        while (slow < nums.length){
            nums[slow] = 0;
            slow++;
        }
    }
}

File: cpp/283-MoveZeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int p1 = 0, p2 = 0, len = nums.size();
        while (p1 < len && p2 < len) {
            //p1: first 0
            while (p1 < len && nums[p1] != 0) {
                p1++;
            }
            p2 = p1;
            //p2: first num != 0
            while (p2 < len && nums[p2] == 0) {
                p2++;
            }
            if (p1 >= len || p2 >= len) {
                return;
            }
            int tmp = nums[p1];
            nums[p1] = nums[p2];
            nums[p2] = tmp;
        }
    }
};
